packages/library-base/src/commonMain/kotlin/io/realm/query/RealmQuery.kt

```diff
@@ -248,15 +253,42 @@ interface RealmSingleQuery<T : RealmObject> {
      * Finds the first object that fulfills the query conditions and returns it asynchronously as a
      * [Flow].
      *
-     * If there is any changes to the object represented by the query, the flow will emit the
-     * updated object. The flow will continue running indefinitely until canceled.
+     * If there is any changes to the first object represented by the query, the flow will emit an
+     * [QueryObjectChange] event depending on the query state. The flow will continue running indefinitely until
+     * canceled.
+     *
+     * If subscribed on an empty query the flow will emit a [PendingObject] event to signal the query
+     * is empty, it would then yield an [InitialObject] event for the first element. On a non-empty
+     * list it would start emitting an [InitialObject] event for its first element.
+     *
+     * Once subscribed and the [InitialObject] event is observed, sequential [UpdatedObject] instances
+     * would be observed if the first element is modified. If the element is deleted a [DeletedObject]
+     * would be yield.
+     *
+     * If the first element is replaced with a new value, an [InitialObject] would be yield for the new
+     * head, and would be follow with [UpdatedObject] on all its changes.
+     *
+     * The state machine, based on if the list has or not a head, looks like this:
+     *
+     *                ┌───────┐
+     *                │ start ├───────────────────────────────┐
+     *                └───┬───┘       [InitialObject]         │
+     *                    │                                   │
+     *    [PendingObject] │                                   │
+     *                    │                                   │
+     *              ┌─────▼─────┐     [InitialObject]    ┌────▼───┐
+     *              │           ├────────────────────────►        ├─────┐[InitialObject]
+     *              │  No Head  │                        │  Head  │     │
+     *              │           ◄────────────────────────┤        │◄────┘[UpdatedObject]
+     *              └───────────┘     [DeletedObject]    └────────┘
```

Comment:
```suggestion
     * If subscribed on an empty query the flow will emit a [PendingObject] event to signal the query
     * is empty, it would then yield an [InitialObject] event for the first element. On a non-empty
     * query result it would start emitting an [InitialObject] event for its first element.
     *
     * Once subscribed and the [InitialObject] event is observed, sequential [UpdatedObject] instances
     * would be observed if the first element is modified. If the element is deleted a [DeletedObject]
     * would be observed.
     *
     * If the first element is replaced with a new object, an [InitialObject] would be sent for the new
     * head, which would be followed by [UpdatedObject] on all subsequent changes to that object.
     *
     * The state machine looks like this:
     *
     *                ┌───────┐
     *                │ start ├───────────────────────────────┐
     *                └───┬───┘       [InitialObject]         │
     *                    │                                   │
     *    [PendingObject] │                                   │
     *                    │                                   │
     *              ┌─────▼─────┐     [InitialObject]    ┌────▼───┐
     *              │           ├────────────────────────►        ├─────┐[InitialObject]
     *              │  No Head  │                        │  Head  │     │
     *              │           ◄────────────────────────┤        │◄────┘[UpdatedObject]
     *              └───────────┘     [DeletedObject]    └────────┘
```

Reply:
Having the state machine in the doc is really awesome, but I'm getting very confused that you have the states as transitions in the diagram instead of them being states... IMO we shouldn't emphasize the underlying implementation that much but focus on the visible user-facing behavior.